File: next-tauri/src/platform/contracts/codeRuntime.ts
export type CodeRuntimeId = 'java' | 'groovy' | 'python' | 'node'
export interface CodeRuntimeStatus { id: CodeRuntimeId; available: boolean; command: string; version: string }
export interface CodeRunSpec { requestId: string; runtime: CodeRuntimeId; code: string; timeoutMs: number; arguments: string[]; workingDirectory: string }
export interface CodeOutputEvent { stream: 'stdout' | 'stderr'; text: string }
export interface CodeRunResult { exitCode: number | null; stdout: string; stderr: string; durationMs: number; command: string; timedOut: boolean; cancelled: boolean }
export interface CodeRuntimeApi { detect(): Promise<CodeRuntimeStatus[]>; run(spec: CodeRunSpec, onOutput: (event: CodeOutputEvent) => void): Promise<CodeRunResult>; cancel(requestId: string): Promise<boolean> }

File: next-tauri/src-tauri/src/contracts/code_runtime.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum CodeRuntimeId {
    Java,
    Groovy,
    Python,
    Node,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeRuntimeStatus {
    pub id: CodeRuntimeId,
    pub available: bool,
    pub command: String,
    pub version: String,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeRunSpec {
    pub request_id: String,
    pub runtime: CodeRuntimeId,
    pub code: String,
    pub timeout_ms: u64,
    pub arguments: Vec<String>,
    pub working_directory: String,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeOutputEvent {
    pub stream: String,
    pub text: String,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeRunResult {
    pub exit_code: Option<i32>,
    pub stdout: String,
    pub stderr: String,
    pub duration_ms: u128,
    pub command: String,
    pub timed_out: bool,
    pub cancelled: bool,
}
